File: motkarta/scoring.py
from __future__ import annotations

from dataclasses import dataclass, field
from math import exp, log, log1p


def logarithmic_count_score(count: float, high_watermark: float = 10000) -> float:
    """Use logarithmic transformation log1p(count) to reduce unfair advantage of places with thousands of reviews."""
    if high_watermark <= 0:
        return 0
    return clamp((log1p(max(0, count)) / log1p(high_watermark)) * 100)


Confidence = str


@dataclass(frozen=True)
class EvidenceSignals:
    specialist_guide: float = 0
    independent_editorial: float = 0
    verified_user_rating: float = 0
    repeat_visits: float = 0
    recent_reviews: float = 0
    credible_reviewers: float = 0
    inspection_status: float = 60
    verified_attributes: float = 0
    data_freshness: float = 0
    confidence: Confidence = "Low"


@dataclass(frozen=True)
class EngagementSignals:
    search_impressions: int = 0
    profile_views: int = 0
    map_marker_clicks: int = 0
    saves: int = 0
    direction_requests: int = 0
    confirmed_visits: int = 0
    repeat_visits: int = 0
    recommendations: int = 0
    recent_saves: int = 0


@dataclass(frozen=True)
class SpecialtyAttributes:
    specialty_verified: bool = False
    own_roastery: bool = False
    traceable_coffee: bool = False
    filter_coffee: bool = False
    espresso_based: bool = False
    rotating_roasters: bool = False
    single_origin: bool = False
    manual_brew_methods: list[str] = field(default_factory=list)
    decaf_available: bool = False
    beans_for_sale: bool = False
    verification_sources: int = 0


@dataclass(frozen=True)
class PlaceInput:
    id: int
    name: str
    kind: str
    area: str
    tags: list[str]
    rating_average: float = 4.1
    reliable_rating_count: int = 0
    review_count: int = 0
    category_mean_rating: float = 4.1
    price_level: int = 2
    mainstream_exposure: float = 0
    days_since_fresh_evidence: int = 365
    evidence: EvidenceSignals = field(default_factory=EvidenceSignals)
    engagement: EngagementSignals = field(default_factory=EngagementSignals)
    specialty: SpecialtyAttributes | None = None


def clamp(value: float, minimum: float = 0, maximum: float = 100) -> float:
    return min(maximum, max(minimum, value))


def bayesian_rating(rating: float, rating_count: int, global_mean: float = 4.1, minimum: int = 30) -> float:
    if rating_count <= 0:
        return global_mean
    return rating_count / (rating_count + minimum) * rating + minimum / (rating_count + minimum) * global_mean


def recency_weight(days_old: int, half_life: int = 180) -> float:
    return exp(-log(2) * max(0, days_old) / half_life)


def bayesian_rate(positive_signals: int, total_signals: int, prior_rate: float = 0.08, prior_weight: int = 50) -> float:
    if total_signals <= 0:
        return prior_rate
    return (positive_signals + prior_rate * prior_weight) / (total_signals + prior_weight)


def verify_specialty_coffee_eligibility(place: PlaceInput) -> bool:
    """A place cannot receive the 'Specialty coffee' label merely from marketing text (e.g. 'premium coffee').

    Requires at least one of four verification gates:
    1. Verification by a recognised specialty guide (specialist_guide > 0)
    2. Verification by editorial team (specialty_verified is True)
    3. Consistent community submissions (verification_sources >= 2)
    4. Sufficient structured evidence from menu/website (structured_signals >= 3)
    """
    guide_verified = place.evidence.specialist_guide > 0
    if not place.specialty:
        return guide_verified

    attributes = place.specialty
    structured_signals = sum(
        [
            attributes.specialty_verified,
            attributes.own_roastery,
            attributes.traceable_coffee,
            attributes.single_origin,
            attributes.rotating_roasters,
            bool(attributes.manual_brew_methods),
            attributes.beans_for_sale,
        ]
    )

    editorial_verified = attributes.specialty_verified
    community_verified = attributes.verification_sources >= 2
    menu_verified = structured_signals >= 3

    return guide_verified or editorial_verified or community_verified or menu_verified


def specialty_confidence(place: PlaceInput) -> float:
    if place.kind != "Specialty coffee" or not verify_specialty_coffee_eligibility(place):
        return 0
    if not place.specialty:
        return 50 if place.evidence.specialist_guide > 0 else 0
    attributes = place.specialty
    structured_signals = sum(
        [
            attributes.specialty_verified,
            attributes.own_roastery,
            attributes.traceable_coffee,
            attributes.filter_coffee,
            attributes.espresso_based,
            attributes.rotating_roasters,
            attributes.single_origin,
            attributes.decaf_available,
            attributes.beans_for_sale,
            bool(attributes.manual_brew_methods),
        ]
    )
    verification_score = min(3, attributes.verification_sources) / 3
    return clamp(structured_signals * 7 + verification_score * 30)


def quality_score(place: PlaceInput) -> float:
    evidence = place.evidence
    base = (
        0.16 * evidence.specialist_guide
        + 0.14 * evidence.independent_editorial
        + 0.14 * evidence.verified_user_rating
        + 0.1 * evidence.repeat_visits
        + 0.1 * evidence.recent_reviews
        + 0.1 * evidence.credible_reviewers
        + 0.06 * evidence.inspection_status
        + 0.1 * evidence.verified_attributes
        + 0.1 * evidence.data_freshness
    )
    return clamp(base + (specialty_confidence(place) * 0.12 if place.kind == "Specialty coffee" else 0))


def freshness_score(place: PlaceInput) -> float:
    return clamp(recency_weight(place.days_since_fresh_evidence) * 100)


def popularity_score(place: PlaceInput) -> dict[str, float]:
    bayesian_user_rating = (
        (bayesian_rating(place.rating_average, place.reliable_rating_count, place.category_mean_rating) - 1) / 4
    ) * 100
    engagement = place.engagement
    exposure_adjusted_engagement = (
        bayesian_rate(
            engagement.saves + engagement.confirmed_visits + engagement.direction_requests,
            max(1, engagement.search_impressions),
            0.08,
            120,
        )
        * 100
    )
    repeat_visit_rate = bayesian_rate(engagement.repeat_visits, max(1, engagement.confirmed_visits), 0.18, 30) * 100
    recent_save_rate = bayesian_rate(engagement.recent_saves, max(1, engagement.saves), 0.38, 25) * 100
    confidence_weight = {"Low": 0.45, "Medium": 0.72, "High": 1}.get(place.evidence.confidence, 0.45)
    source_consensus = clamp(
        (
            place.evidence.specialist_guide
            + place.evidence.independent_editorial
            + place.evidence.verified_user_rating
            + place.evidence.credible_reviewers
        )
        / 4
        * 100
        * confidence_weight
    )
    score = clamp(
        0.3 * bayesian_user_rating
        + 0.25 * exposure_adjusted_engagement
        + 0.2 * repeat_visit_rate
        + 0.15 * recent_save_rate
        + 0.1 * source_consensus
    )
    return {
        "score": score,
        "bayesian_user_rating": bayesian_user_rating,
        "exposure_adjusted_engagement": exposure_adjusted_engagement,
        "repeat_visit_rate": repeat_visit_rate,
        "recent_save_rate": recent_save_rate,
        "source_consensus": source_consensus,
    }


def relevance_score(place: PlaceInput) -> float:
    return 50


def discovery_score(place: PlaceInput, quality: float) -> dict[str, float]:
    specialist_confidence = place.evidence.specialist_guide * 70 + specialty_confidence(place) * 0.3
    local_engagement = (
        bayesian_rate(
            place.engagement.map_marker_clicks + place.engagement.saves,
            max(1, place.engagement.search_impressions),
            0.1,
            80,
        )
        * 100
    )
    score = clamp(
        0.4 * quality
        + 0.25 * specialist_confidence
        + 0.2 * local_engagement
        + 0.15 * freshness_score(place)
        - 0.25 * place.mainstream_exposure
    )
    return {"score": score, "specialist_confidence": clamp(specialist_confidence), "local_engagement": local_engagement}


def score_place(place: PlaceInput) -> dict[str, float]:
    quality = quality_score(place)
    popularity = popularity_score(place)
    relevance = relevance_score(place)
    discovery = discovery_score(place, quality)
    freshness = freshness_score(place)
    recommendation = clamp(
        0.35 * relevance + 0.25 * quality + 0.15 * popularity["score"] + 0.15 * discovery["score"] + 0.1 * freshness
    )
    return {
        "quality": quality,
        "popularity": popularity["score"],
        "relevance": relevance,
        "discovery": discovery["score"],
        "freshness": freshness,
        "recommendation": recommendation,
    }

File: lib/scoring.ts
export type EstablishmentType =
  | "Restaurant"
  | "Bakery"
  | "Café"
  | "Specialty coffee";

export type Confidence = "Low" | "Medium" | "High";
export type PlaceLifecycleState = "baseline" | "candidate" | "verified" | "featured";

export type SpecialtyAttributes = {
  specialtyVerified: boolean;
  ownRoastery: boolean;
  traceableCoffee: boolean;
  filterCoffee: boolean;
  espressoBased: boolean;
  rotatingRoasters: boolean;
  singleOrigin: boolean;
  manualBrewMethods: string[];
  decafAvailable: boolean;
  beansForSale: boolean;
  verificationSources: number;
};

export type EvidenceSignals = {
  specialistGuide: number;
  independentEditorial: number;
  verifiedUserRating: number;
  repeatVisits: number;
  recentReviews: number;
  credibleReviewers: number;
  inspectionStatus: number;
  verifiedAttributes: number;
  dataFreshness: number;
  confidence: Confidence;
};

export type EngagementSignals = {
  searchImpressions: number;
  profileViews: number;
  mapMarkerClicks: number;
  saves: number;
  directionRequests: number;
  confirmedVisits: number;
  repeatVisits: number;
  recommendations: number;
  recentSaves: number;
};

export type UserPreferences = {
  kind?: EstablishmentType;
  tags?: string[];
  priceLevel?: number;
  district?: string;
  purpose?: "breakfast" | "fika" | "lunch" | "dinner" | "work";
  independentOnly?: boolean;
};

export type PlaceInput = {
  id: number;
  name: string;
  kind: EstablishmentType;
  cuisine?: string;
  area: string;
  address?: string;
  note: string;
  tags: string[];
  discoveryReasons?: string[];
  discoverySignals?: Record<string, boolean>;
  sourceName?: string;
  lastUpdated?: string;
  evidenceLabel: string;
  ratingAverage: number;
  reliableRatingCount: number;
  reviewCount: number;
  categoryMeanRating: number;
  categoryPopularityRaw: number;
  localPopularityPercentile: number;
  priceLevel: number;
  mainstreamExposure: number;
  ageDays: number;
  daysSinceFreshEvidence: number;
  is_hidden_gem?: boolean;
  lifecycleState?: PlaceLifecycleState;
  validationLabel?: "known_mainstream" | "known_hidden_gem" | "not_enough_evidence" | "closed_wrong_category";
  validationNotes?: string;
  evidence: EvidenceSignals;
  engagement: EngagementSignals;
  specialty?: SpecialtyAttributes;
  latitude?: number;
  longitude?: number;
  website?: string;
  x: number;
  y: number;
};

export type ScoreBreakdown = {
  quality: number;
  popularity: number;
  relevance: number;
  discovery: number;
  freshness: number;
  recommendation: number;
  bayesianUserRating: number;
  exposureAdjustedEngagement: number;
  repeatVisitRate: number;
  recentSaveRate: number;
  crossSourceConsensus: number;
  specialistConfidence: number;
  localEngagement: number;
};

export type VerificationSourceStatus = {
  verified: boolean;
  label: string;
  detail: string;
};

export type VerificationBreakdown = {
  specialistGuide: VerificationSourceStatus;
  editorialTeam: VerificationSourceStatus;
  communitySubmissions: VerificationSourceStatus;
  structuredEvidence: VerificationSourceStatus;
  verifiedSourcesCount: number;
  confidenceLevel: Confidence;
  confidenceScore: number;
  summary: string;
};

export type ScoredPlace = PlaceInput & {
  scores: ScoreBreakdown;
  verification: VerificationBreakdown;
  hiddenGem: HiddenGemEligibility;
};

export type HiddenGemGateStatus = {
  passed: boolean;
  label: string;
  detail: string;
};

export type HiddenGemEligibility = {
  eligible: boolean;
  independentEvidenceCount: number;
  gates: {
    lowMainstreamExposure: HiddenGemGateStatus;
    independentEvidence: HiddenGemGateStatus;
    currentExistence: HiddenGemGateStatus;
    distinctiveness: HiddenGemGateStatus;
    lifecycle: HiddenGemGateStatus;
  };
};

const confidenceWeight: Record<Confidence, number> = {
  Low: 0.45,
  Medium: 0.72,
  High: 1,
};

const defaultEvidence: EvidenceSignals = {
  specialistGuide: 0,
  independentEditorial: 0,
  verifiedUserRating: 0,
  repeatVisits: 0,
  recentReviews: 0,
  credibleReviewers: 0,
  inspectionStatus: 0,
  verifiedAttributes: 0,
  dataFreshness: 0,
  confidence: "Low",
};

const defaultEngagement: EngagementSignals = {
  searchImpressions: 0,
  profileViews: 0,
  mapMarkerClicks: 0,
  saves: 0,
  directionRequests: 0,
  confirmedVisits: 0,
  repeatVisits: 0,
  recommendations: 0,
  recentSaves: 0,
};

function finiteNumber(value: unknown, fallback = 0) {
  return typeof value === "number" && Number.isFinite(value) ? value : fallback;
}

function normalizedConfidence(value: unknown): Confidence {
  return value === "High" || value === "Medium" || value === "Low" ? value : "Low";
}

function normalizeEvidence(evidence: Partial<EvidenceSignals> | undefined): EvidenceSignals {
  const source = { ...defaultEvidence, ...evidence };

  return {
    specialistGuide: finiteNumber(source.specialistGuide),
    independentEditorial: finiteNumber(source.independentEditorial),
    verifiedUserRating: finiteNumber(source.verifiedUserRating),
    repeatVisits: finiteNumber(source.repeatVisits),
    recentReviews: finiteNumber(source.recentReviews),
    credibleReviewers: finiteNumber(source.credibleReviewers),
    inspectionStatus: finiteNumber(source.inspectionStatus),
    verifiedAttributes: finiteNumber(source.verifiedAttributes),
    dataFreshness: finiteNumber(source.dataFreshness),
    confidence: normalizedConfidence(source.confidence),
  };
}

function normalizeEngagement(engagement: Partial<EngagementSignals> | undefined): EngagementSignals {
  const source = { ...defaultEngagement, ...engagement };

  return {
    searchImpressions: finiteNumber(source.searchImpressions),
    profileViews: finiteNumber(source.profileViews),
    mapMarkerClicks: finiteNumber(source.mapMarkerClicks),
    saves: finiteNumber(source.saves),
    directionRequests: finiteNumber(source.directionRequests),
    confirmedVisits: finiteNumber(source.confirmedVisits),
    repeatVisits: finiteNumber(source.repeatVisits),
    recommendations: finiteNumber(source.recommendations),
    recentSaves: finiteNumber(source.recentSaves),
  };
}

export function clamp(value: number, min = 0, max = 100) {
  if (!Number.isFinite(value)) {
    return min;
  }

  return Math.min(max, Math.max(min, value));
}

export function bayesianRating(
  rating: number,
  ratingCount: number,
  categoryMean = 4.1,
  minimumEvidence = 30,
) {
  const safeCategoryMean = finiteNumber(categoryMean, 4.1);
  const safeRatingCount = Math.max(0, finiteNumber(ratingCount));
  const safeMinimumEvidence = Math.max(0, finiteNumber(minimumEvidence, 30));

  if (safeRatingCount <= 0) {
    return safeCategoryMean;
  }

  const safeRating = finiteNumber(rating, safeCategoryMean);
  const totalEvidence = safeRatingCount + safeMinimumEvidence;

  return (
    (safeRatingCount / totalEvidence) * safeRating +
    (safeMinimumEvidence / totalEvidence) * safeCategoryMean
  );
}

export function recencyWeight(daysOld: number, halfLife = 180) {
  const safeHalfLife = finiteNumber(halfLife, 180);
  if (safeHalfLife <= 0) {
    return 0;
  }

  return Math.exp((-Math.log(2) * Math.max(0, finiteNumber(daysOld))) / safeHalfLife);
}

export function bayesianRate(
  positiveSignals: number,
  totalSignals: number,
  priorRate = 0.08,
  priorWeight = 50,
) {
  const safePositiveSignals = Math.max(0, finiteNumber(positiveSignals));
  const safeTotalSignals = Math.max(0, finiteNumber(totalSignals));
  const safePriorRate = finiteNumber(priorRate, 0.08);
  const safePriorWeight = Math.max(0, finiteNumber(priorWeight, 50));

  if (safeTotalSignals <= 0) {
    return safePriorRate;
  }

  return (safePositiveSignals + safePriorRate * safePriorWeight) / (safeTotalSignals + safePriorWeight);
}

export function logarithmicCountScore(count: number, highWatermark: number) {
  const safeHighWatermark = finiteNumber(highWatermark);
  if (safeHighWatermark <= 0) {
    return 0;
  }

  return clamp((Math.log1p(Math.max(0, finiteNumber(count))) / Math.log1p(safeHighWatermark)) * 100);
}

function sourceConsensus(evidence: EvidenceSignals | Partial<EvidenceSignals> | undefined) {
  const normalized = normalizeEvidence(evidence);
  const sourceCount =
    normalized.specialistGuide +
    normalized.independentEditorial +
    normalized.verifiedUserRating +
    normalized.credibleReviewers;

  return clamp((sourceCount / 4) * 100 * confidenceWeight[normalized.confidence]);
}

function signalStrength(value: number | undefined) {
  const raw = finiteNumber(value);
  return clamp(raw <= 1 ? raw * 100 : raw);
}

export function independentEvidenceCount(place: PlaceInput) {
  const evidence = normalizeEvidence(place.evidence);
  const signals = [
    signalStrength(evidence.specialistGuide) >= 40,
    signalStrength(evidence.independentEditorial) >= 40,
    signalStrength(evidence.verifiedUserRating) >= 40 ||
      signalStrength(evidence.credibleReviewers) >= 40 ||
      (place.reliableRatingCount ?? 0) >= 20,
    signalStrength(evidence.inspectionStatus) >= 80,
    signalStrength(evidence.verifiedAttributes) >= 50 ||
      (place.specialty?.verificationSources ?? 0) >= 2 ||
      place.specialty?.specialtyVerified === true,
  ];

  return signals.filter(Boolean).length;
}

export function isUserVisibleLifecycleState(state: PlaceLifecycleState | undefined) {
  return state !== "candidate";
}

export function evaluateHiddenGemGates(place: PlaceInput): HiddenGemEligibility {
  const evidence = normalizeEvidence(place.evidence);
  const evidenceCount = independentEvidenceCount(place);
  const nonGenericTags = (place.tags ?? []).filter((tag) => {
    const normalized = tag.toLowerCase().trim();
    return ![
      "restaurant",
      "bakery",
      "café",
      "cafe",
      "specialty coffee",
      "independent",
      "community submission",
      "pending verification",
      "openstreetmap",
      "osm",
    ].includes(normalized);
  });
  const hasDistinctiveStructuredSignal =
    Boolean(place.discoveryReasons?.length) ||
    Object.values(place.discoverySignals ?? {}).some(Boolean) ||
    nonGenericTags.length >= 2 ||
    (place.cuisine ? !["general", "restaurant", "cafe", "coffee"].includes(place.cuisine.toLowerCase()) : false) ||
    (place.kind === "Specialty coffee" && verifySpecialtyCoffeeEligibility(place));

  const lifecycleState = place.lifecycleState ?? "baseline";
  const isKnownBadValidation = place.validationLabel === "closed_wrong_category";
  const currentExistence =
    !isKnownBadValidation &&
    (finiteNumber(place.daysSinceFreshEvidence, 365) <= 365 ||
      signalStrength(evidence.dataFreshness) >= 50 ||
      signalStrength(evidence.inspectionStatus) >= 80);

  const gates = {
    lowMainstreamExposure: {
      passed: finiteNumber(place.mainstreamExposure, 100) <= 40,
      label: "Low mainstream exposure",
      detail: `mainstreamExposure ${finiteNumber(place.mainstreamExposure, 100).toFixed(0)} <= 40`,
    },
    independentEvidence: {
      passed: evidenceCount >= 2,
      label: "Two independent evidence signals",
      detail: `${evidenceCount} independent signal${evidenceCount === 1 ? "" : "s"} found`,
    },
    currentExistence: {
      passed: currentExistence,
      label: "Current existence",
      detail: currentExistence ? "Recent or municipal/field evidence is present" : "Needs recent existence verification",
    },
    distinctiveness: {
      passed: hasDistinctiveStructuredSignal,
      label: "Distinctiveness",
      detail: hasDistinctiveStructuredSignal ? "Distinctive tags, source reasons, or specialty attributes present" : "Needs a specific reason beyond obscurity",
    },
    lifecycle: {
      passed: isUserVisibleLifecycleState(lifecycleState) && !isKnownBadValidation,
      label: "Visible lifecycle state",
      detail: `state=${lifecycleState}${isKnownBadValidation ? ", validation=closed_wrong_category" : ""}`,
    },
  };

  return {
    eligible: Object.values(gates).every((gate) => gate.passed),
    independentEvidenceCount: evidenceCount,
    gates,
  };
}

export function verifySpecialtyCoffeeEligibility(place: PlaceInput): boolean {
  const guideVerified = (place.evidence?.specialistGuide ?? 0) > 0;
  if (!place.specialty) {
    return guideVerified;
  }

  const attributes = place.specialty;
  const structuredSignals = [
    attributes.specialtyVerified,
    attributes.ownRoastery,
    attributes.traceableCoffee,
    attributes.singleOrigin,
    attributes.rotatingRoasters,
    Boolean(attributes.manualBrewMethods?.length),
    attributes.beansForSale,
  ].filter(Boolean).length;

  const editorialVerified = attributes.specialtyVerified;
  const communityVerified = (attributes.verificationSources ?? 0) >= 2;
  const menuVerified = structuredSignals >= 3;

  return guideVerified || editorialVerified || communityVerified || menuVerified;
}

function specialtyConfidence(place: PlaceInput) {
  if (place.kind !== "Specialty coffee" || !verifySpecialtyCoffeeEligibility(place)) {
    return 0;
  }

  if (!place.specialty) {
    return (place.evidence?.specialistGuide ?? 0) > 0 ? 50 : 0;
  }

  const attributes = place.specialty;
  const structuredSignals = [
    attributes.specialtyVerified,
    attributes.ownRoastery,
    attributes.traceableCoffee,
    attributes.filterCoffee,
    attributes.espressoBased,
    attributes.rotatingRoasters,
    attributes.singleOrigin,
    attributes.decafAvailable,
    attributes.beansForSale,
    (attributes.manualBrewMethods?.length ?? 0) > 0,
  ].filter(Boolean).length;

  const verificationScore = Math.min(3, finiteNumber(attributes.verificationSources)) / 3;
  return clamp(structuredSignals * 7 + verificationScore * 30);
}

export function qualityScore(place: PlaceInput) {
  const evidence = normalizeEvidence(place.evidence);
  const base =
    0.16 * evidence.specialistGuide +
    0.14 * evidence.independentEditorial +
    0.14 * evidence.verifiedUserRating +
    0.1 * evidence.repeatVisits +
    0.1 * evidence.recentReviews +
    0.1 * evidence.credibleReviewers +
    0.06 * evidence.inspectionStatus +
    0.1 * evidence.verifiedAttributes +
    0.1 * evidence.dataFreshness;

  const specialtyBonus = place.kind === "Specialty coffee" ? specialtyConfidence(place) * 0.12 : 0;
  return clamp(base + specialtyBonus);
}

export function freshnessScore(place: PlaceInput) {
  return clamp(recencyWeight(place.daysSinceFreshEvidence) * 100);
}

export function popularityScore(place: PlaceInput) {
  const eng = normalizeEngagement(place.engagement);

  const bayesianUserRating =
    ((bayesianRating(
      place.ratingAverage,
      place.reliableRatingCount,
      place.categoryMeanRating,
    ) - 1) /
      4) *
    100;

  const exposureAdjustedEngagement =
    bayesianRate(
      eng.saves + eng.confirmedVisits + eng.directionRequests,
      Math.max(1, eng.searchImpressions),
      0.08,
      120,
    ) * 100;

  const repeatVisitRate =
    bayesianRate(eng.repeatVisits, Math.max(1, eng.confirmedVisits), 0.18, 30) * 100;
  const recentSaveRate =
    bayesianRate(eng.recentSaves, Math.max(1, eng.saves), 0.38, 25) * 100;

  const crossSourceConsensus = sourceConsensus(place.evidence);

  return {
    score: clamp(
      0.3 * bayesianUserRating +
        0.25 * exposureAdjustedEngagement +
        0.2 * repeatVisitRate +
        0.15 * recentSaveRate +
        0.1 * crossSourceConsensus,
    ),
    bayesianUserRating,
    exposureAdjustedEngagement,
    repeatVisitRate,
    recentSaveRate,
    crossSourceConsensus,
  };
}

export function relevanceScore(place: PlaceInput, preferences: UserPreferences = {}) {
  let score = 50;

  if (preferences.kind) {
    score += place.kind === preferences.kind ? 22 : -12;
  }

  if (preferences.district) {
    score += place.area === preferences.district ? 12 : 0;
  }

  if (typeof preferences.priceLevel === "number") {
    score += Math.max(0, 12 - Math.abs(finiteNumber(place.priceLevel) - preferences.priceLevel) * 6);
  }

  if (preferences.independentOnly) {
    score += (place.tags ?? []).includes("Independent") ? 10 : -10;
  }

  if (preferences.tags?.length) {
    const lowerTags = (place.tags ?? []).map((tag) => tag.toLowerCase());
    const matches = preferences.tags.filter((tag) =>
      lowerTags.some((placeTag) => placeTag.includes(tag.toLowerCase())),
    );
    score += Math.min(24, matches.length * 8);
  }

  if (preferences.purpose && (place.tags ?? []).some((tag) => tag.toLowerCase() === preferences.purpose)) {
    score += 8;
  }

  return clamp(score);
}

export function discoveryScore(place: PlaceInput, quality: number) {
  const eng = normalizeEngagement(place.engagement);

  const evidence = normalizeEvidence(place.evidence);
  const specialistConfidence = evidence.specialistGuide * 70 + specialtyConfidence(place) * 0.3;
  const localEngagement =
    bayesianRate(
      eng.mapMarkerClicks + eng.saves,
      Math.max(1, eng.searchImpressions),
      0.1,
      80,
    ) * 100;
  const dataFreshness = freshnessScore(place);

  return {
    score: clamp(
      0.4 * quality +
        0.25 * specialistConfidence +
        0.2 * localEngagement +
        0.15 * dataFreshness -
        0.25 * finiteNumber(place.mainstreamExposure, 100),
    ),
    specialistConfidence: clamp(specialistConfidence),
    localEngagement,
  };
}

export function computeVerificationBreakdown(place: PlaceInput): VerificationBreakdown {
  const evidence = normalizeEvidence(place.evidence);
  const tags = place.tags ?? [];
  const guideVerified =
    evidence.specialistGuide >= 0.4 ||
    (place.evidenceLabel?.toLowerCase().includes("guide") ?? false) ||
    (place.evidenceLabel?.toLowerCase().includes("specialist") ?? false) ||
    tags.some((t) => ["white guide", "michelin", "star wine list", "specialist guide", "sca", "craft guide"].includes(t.toLowerCase()));

  const editorialVerified =
    evidence.independentEditorial >= 0.3 ||
    (place.evidenceLabel?.toLowerCase().includes("editorial") ?? false) ||
    (place.evidenceLabel?.toLowerCase().includes("official site") ?? false) ||
    (place.evidenceLabel?.toLowerCase().includes("osm") ?? false) ||
    Boolean(place.sourceName) ||
    (place.specialty?.specialtyVerified === true);

  const communityCount =
    finiteNumber(place.specialty?.verificationSources) +
    Math.floor(evidence.repeatVisits / 10) +
    Math.floor(finiteNumber(place.engagement?.confirmedVisits) / 15) +
    (finiteNumber(place.localPopularityPercentile) > 0.4 ? 2 : 1);

  const communityVerified =
    communityCount >= 2 ||
    (place.reliableRatingCount ?? 0) >= 20 ||
    (place.engagement?.saves ?? 0) > 10;

  const structuredSignals = [
    Boolean(place.website),
    Boolean(place.address),
    Boolean(place.area && place.area !== "Stockholm"),
    Boolean(tags.length),
    place.specialty?.traceableCoffee ?? false,
    place.specialty?.filterCoffee ?? false,
    place.specialty?.singleOrigin ?? false,
    place.specialty?.ownRoastery ?? false,
    (place.specialty?.manualBrewMethods?.length ?? 0) > 0,
  ].filter(Boolean).length;

  const structuredVerified =
    structuredSignals >= 2 ||
    Boolean(place.website) ||
    evidence.verifiedAttributes >= 30;

  const verifiedSourcesCount = [guideVerified, editorialVerified, communityVerified, structuredVerified].filter(Boolean).length;

  const confidenceScore = clamp(
    verifiedSourcesCount * 25 +
      (evidence.confidence === "High" ? 15 : evidence.confidence === "Medium" ? 5 : 0),
  );

  const confidenceLevel: Confidence =
    verifiedSourcesCount >= 3 ? "High" : verifiedSourcesCount >= 2 ? "Medium" : "Low";

  const verifiedNames = [
    guideVerified && "Specialty Guide",
    editorialVerified && "Editorial Team",
    communityVerified && "Community Consensus",
    structuredVerified && "Structured Menu & Web",
  ].filter(Boolean) as string[];

  const summary = verifiedSourcesCount >= 3
    ? `Verified by ${verifiedSourcesCount} independent sources (${verifiedNames.join(", ")})`
    : verifiedSourcesCount === 2
    ? `Verified by 2 independent sources (${verifiedNames.join(", ")})`
    : `Verified by 1 source (${verifiedNames[0] ?? "Open Data Baseline"})`;

  return {
    specialistGuide: {
      verified: guideVerified,
      label: "Recognised Specialty Guide",
      detail: guideVerified ? "Verified by White Guide / Specialist Guide" : "Unverified by specialty guide",
    },
    editorialTeam: {
      verified: editorialVerified,
      label: "Editorial Team & Admin Audit",
      detail: editorialVerified ? "Audited & verified by editorial team" : "Pending editorial audit",
    },
    communitySubmissions: {
      verified: communityVerified,
      label: "Consistent Community Submissions",
      detail: communityVerified ? `Verified by ${communityCount || 3}+ consistent community submissions` : "Awaiting community submissions",
    },
    structuredEvidence: {
      verified: structuredVerified,
      label: "Structured Menu & Web Evidence",
      detail: structuredVerified ? "Menu, website & operational signals verified" : "Requires live menu verification",
    },
    verifiedSourcesCount,
    confidenceLevel,
    confidenceScore,
    summary,
  };
}

export function scorePlace(place: PlaceInput, preferences: UserPreferences = {}): ScoredPlace {
  const quality = qualityScore(place);
  const popularity = popularityScore(place);
  const relevance = relevanceScore(place, preferences);
  const discovery = discoveryScore(place, quality);
  const freshness = freshnessScore(place);
  const verification = computeVerificationBreakdown(place);
  const hiddenGem = evaluateHiddenGemGates(place);
  const recommendation = clamp(
    0.35 * relevance +
      0.25 * quality +
      0.15 * popularity.score +
      0.15 * discovery.score +
      0.1 * freshness,
  );

  return {
    ...place,
    is_hidden_gem: Boolean(place.is_hidden_gem && hiddenGem.eligible),
    scores: {
      quality,
      popularity: popularity.score,
      relevance,
      discovery: discovery.score,
      freshness,
      recommendation,
      bayesianUserRating: popularity.bayesianUserRating,
      exposureAdjustedEngagement: popularity.exposureAdjustedEngagement,
      repeatVisitRate: popularity.repeatVisitRate,
      recentSaveRate: popularity.recentSaveRate,
      crossSourceConsensus: popularity.crossSourceConsensus,
      specialistConfidence: discovery.specialistConfidence,
      localEngagement: discovery.localEngagement,
    },
    verification,
    hiddenGem,
  };
}

export function addLocalPopularityPercentiles<T extends PlaceInput>(places: T[]) {
  return places.map((place) => {
    const peers = places
      .filter((peer) => peer.kind === place.kind && peer.area === place.area)
      .sort((a, b) => finiteNumber(a.categoryPopularityRaw) - finiteNumber(b.categoryPopularityRaw));
    const rank = peers.findIndex((peer) => peer.id === place.id) + 1;

    return {
      ...place,
      localPopularityPercentile: peers.length > 1 ? rank / peers.length : 0.5,
    };
  });
}
